graph/run-async.go

```diff
@@ -135,22 +135,60 @@ func ProcessMessages[V VPI[V], E EPI[E], M MVI[M], N any](alg Algorithm[V, E, M,
 	return false, algCount
 }
 
+func (gt *GraphThread[V, E, M, N]) checkCommandsAsync(epoch *bool) {
+	switch <-gt.Command {
+	case BLOCK_ALL:
+		gt.Response <- ACK
+		resp := <-gt.Command // BLOCK and wait for resume
+		if resp != RESUME {
+			log.Panic().Msg("Expected to resume after blocked")
+		}
+	case BLOCK_TOP:
+		log.Panic().Msg("There's no topology changes to block")
+	case RESUME:
+		// No ack needed.
+		break
+	case BLOCK_ALG_IF_TOP:
+		log.Panic().Msg("There's no topology changes")
+	case BSP_SYNC:
+		log.Panic().Msg("Not supported yet")
+	case EPOCH:
+		*epoch = true
+		// Ack after complete.
+	}
+}
+
 // A thread that will process messages until it is done.
 func ConvergeAsyncThread[V VPI[V], E EPI[E], M MVI[M], N any, A Algorithm[V, E, M, N]](alg A, g *Graph[V, E, M, N], tidx uint32, wg *sync.WaitGroup) {
 	runtime.LockOSThread()
 	gt := &g.GraphThreads[tidx]
-	gt.Status = APPLY_MSG
 	algCount := 0
 	algNoCountTimes := 0
+	epoch := false
+
 	for completed := false; !completed; {
-		completed, algCount = ProcessMessages[V, E, M, N](alg, g, gt, true)
+		if len(gt.Command) > 0 {
+			gt.Status = RECV_CMD
+			gt.checkCommandsAsync(&epoch)
+		}
 
-		if algCount == 0 { // Minor back off if we didn't get, and keep getting, no messages.
+		gt.Status = APPLY_MSG
+		completed, algCount = ProcessMessages[V, E, M, N](alg, g, gt, true)
+		if epoch && completed {
+			gt.Status = DONE
+			gt.Response <- ACK
+			resp := <-gt.Command // BLOCK and wait for resume
+			if resp != RESUME {
+				log.Panic().Msg("Expected to resume after blocked")
+			}
+			epoch = false
+			completed = false
+			algNoCountTimes = 0
+		} else if algCount == 0 { // Minor back off if we didn't get, and keep getting, no messages.
 			algNoCountTimes++
 			if algNoCountTimes%100 == 0 {
 				gt.Status = BACKOFF_ALG
 				utils.BackOff(algNoCountTimes / 100)
-				gt.Status = APPLY_MSG
```

Comment:
Hm, note this approach is better because it does not have to write to gt.Status every loop.

Reply:
Oh, it probably doesn't matter now because we need to write `gt.Status = RECV_CMD`.

Reply:
That's a good point :) Updated.

Reply:
Still need this one here, then good to merge!